src/validationinterface.h

```diff
@@ -160,8 +162,13 @@ class ValidationSignals {
 private:
     std::unique_ptr<ValidationSignalsImpl> m_internals;
 
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    std::unique_ptr<util::TaskRunnerInterface> m_serial_task_runner;
```

Comment:
In commit "kernel: Remove dependency on CScheduler" (492ff2c504d1da3f4ae8704a5485a286ee76f0d9)

I'd consider the leaving this declaration inside the ValidationSignalsImpl class instead of moving it here. This would adhere more closely to the [pointer-to-implementation](https://en.cppreference.com/w/cpp/language/pimpl) pattern and ensure future ValidationSignalsImpl methods can access all private members of the class, not just some of them. It would also just move less code around unnecessarily.

Reply:
> ensure future ValidationSignalsImpl methods can access all private members of the class, not just some of them.

I don't follow here. `ValidationSignalsImpl` is not a friend of `ValidationSignals`, so I'm not sure which access you are talking about here. Moving it just seems more SOLID to me and I am not sure how this is violating the pimpl pattern, but I feel like I am missing something.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/28960#discussion_r1459279757

[Pointer-to-implementation](https://en.cppreference.com/w/cpp/language/pimpl) is a c++ specific pattern. It's a simple trick to divide a *single* class into two halves so that the private parts of the class are in the .cpp file rather than the .h file. It doesn't have to do with SOLID principles or object oriented design.

The typical way to define a class in c++ is

```
class MyClass
{
public:
   // MyClass public methods and members
private:
   // MyClass private methods and members
};
```
This works well in most cases, but in some cases you don't want to expose private parts of the class in the header. For example if you are trying to avoid adding private dependencies to the header, or trying to define a stable ABI. In these cases, you can use the pointer to implementation trick to divide `MyClass` into two parts, with the public part in the header file:

```c++
class MyClass
{
public:
   // MyClass public methods only
private:
   std::unique_ptr<MyClassImpl> m_impl;
};
```

And the private part in the cpp file:

```c++
class MyClassImpl
{
       // MyClass private members and methods.
};
```

The problem with f442a3a5b2a0a58bc263fbb9c87e8e4715de103a is that it is not following either of these two approaches, but instead taking a hybrid approach when private state is split up across both the public and private classes. This is bad, because if it means that the private methods only have access to part of the private state, not all of it.

In OO terms, ValidationSignals and ValidationSignalsImpl do not represent separate classes. They just represent the public and private parts of a single class, that are declared in separate halves because of C++ headers weirdness that doesn't exist in other languages.

If you actually think that ValidationSignals and ValidationSignalsImpl *should* be separate OO classes, I think you will want to rename the ValidationSignalsImpl to reflect the actual role of this class rather than taking the name of the class which contains it.